jobs/import_process/worker.go

```diff
@@ -125,51 +124,50 @@ func (w *ImportProcessWorker) doJob(job *model.Job) {
 	}
 	defer importFile.Close()
 
-	// TODO (MM-30187): improve this process by eliminating the need to unzip the import
-	// file locally and instead do the whole bulk import process in memory by
-	// streaming the import file.
-
-	// create a temporary dir to extract the zipped import file.
-	dir, err := ioutil.TempDir("", "import")
+	importZipReader, err := zip.NewReader(importFile.(io.ReaderAt), importFileSize)
 	if err != nil {
-		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.tmp_dir", nil, err.Error(), http.StatusInternalServerError)
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.open_file", nil, err.Error(), http.StatusInternalServerError)
 		w.setJobError(job, appError)
 		return
 	}
-	defer os.RemoveAll(dir)
-
-	// extract the contents of the zipped file.
-	paths, err := utils.UnzipToPath(importFile.(io.ReaderAt), importFileSize, dir)
-	if err != nil {
-		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.unzip", nil, err.Error(), http.StatusInternalServerError)
+	if importZipReader == nil {
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.open_file", nil, "importZipReader was nil", http.StatusInternalServerError)
 		w.setJobError(job, appError)
 		return
 	}
 
 	// find JSONL import file.
-	var jsonFilePath string
-	for _, path := range paths {
-		if filepath.Ext(path) == ".jsonl" {
-			jsonFilePath = path
-			break
+	var jsonFile io.ReadCloser
+	for _, f := range importZipReader.File {
+		if filepath.Ext(f.Name) != ".jsonl" {
+			continue
+		}
+		// avoid "zip slip"
+		if strings.Contains(f.Name, "..") {
+			appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.open_file", nil, "jsonFilePath contained ..", http.StatusBadRequest)
+			w.setJobError(job, appError)
+			return
 		}
-	}
 
-	if jsonFilePath == "" {
-		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.missing_jsonl", nil, "", http.StatusBadRequest)
-		w.setJobError(job, appError)
-		return
+		jsonFile, err = f.Open()
+		if err != nil {
+			appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.open_file", nil, err.Error(), http.StatusInternalServerError)
+			w.setJobError(job, appError)
+			return
+		}
+
+		defer jsonFile.Close()
+		break
 	}
 
-	jsonFile, err := os.Open(jsonFilePath)
-	if err != nil {
-		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.open_file", nil, err.Error(), http.StatusInternalServerError)
+	if jsonFile == nil {
```

Comment:
Ah yes, I see what you mean. That will be cleaner, thanks.

Reply:
What if the file that is opened contains files but doesn't contain the `.jsonl` file? Then if we just check the length of `importZipReader.File` it will be `> 0` however `jsonFile` will be `nil` after the search in the for loop is complete.

Reply:
I see, makes sense then. Can we update the error message as stating that the file does not exist? Because `nil` may not be descriptive for the system admins. It could be something like: "there are no supported file formats in the import file".

Reply:
We can probably resurrect `import_process.worker.do_job.missing_jsonl`

Reply:
Done